designs/default-zero-value.md

```diff
@@ -0,0 +1,857 @@
+   think a structure member is required, but the service has since transitioned
+   the member to use the `@default` trait.
+3. Client implementations SHOULD tolerate structure members marked as
+   `@required` that have no serialized value. For example, if a service
+   migrates a member from `@required` to `@default`, then older clients SHOULD
+   gracefully handle the zero value of the member being omitted on the wire.
+   In this case, rather than failing, a client SHOULD set the member value to
+   its default zero value. Failing to deserialize the structure is a bad
+   outcome because what the service perceived as a backward compatible change
+   (i.e., replacing the `@required` trait with the `@default` trait) could
+   break previously generated clients.
+
+
+## AWS specific changes
+
+### AWS Protocols
+
+All existing AWS protocols (that is, `aws.protocols#restJson1`,
+`aws.protocols#awsJson1_0`, `aws.protocols#awsJson1_1`,
+`aws.protocols#restJson1`, `aws.protocols#restXml`, `aws.protocols#awsQuery`,
+and `aws.protocols#ec2Query`) MUST adhere to the following requirements, and
+all future AWS protocols SHOULD adhere to the following:
+
+1. To avoid information disclosure, serializers SHOULD omit the default zero
+   value of structure members marked with the `@default` trait.
+2. Deserializers MUST tolerate receiving the default zero value of a structure
+   member marked with the `@default` trait.
+3. Client deserializers MUST fill in a default zero value for structure members
+   marked with the `@required` trait that have no serialized value, and the
+   targeted shape supports zero values. This prevents older clients from
+   failing to deserialize structures at runtime when the `@required` trait is
+   replaced with the `@default` trait.
+
+
+### Balancing old and new assumptions
+
+Smithy comes from a long line of service frameworks built inside Amazon dating
+back nearly 20 years. Until this proposal, the required trait was only treated
+as server-side input validation. This causes two major issues that impacts the
+AWS SDK team's ability to fully use the required trait for code generation:
+
+1. Many teams did not consider the required trait as something that impacts
+   output. This means that there are many members in AWS that should be
+   marked as required but are not, and there are members that are required in
+   input, but not always present in output.
+2. Most teams designed their APIs under the guidance that the required trait
+   can be removed without breaking their end users.
+
+We will work to improve the accuracy and consistency of the required trait in
+AWS API models where possible, but it will take significant time and effort
+from hundreds of different teams within AWS. To accommodate this shift in
+modeling changes, we will introduce AWS-specific traits to influence code
+generation and provide AWS SDK specific code generation recommendations.
+
+
+#### `aws.api#clientOptional` trait
```

Comment:
This is a downstream consideration, but generic shape generation will need to be set up to expose the right hooks for this AWS specific update.

Reply:
If nothing else, I know the current generators allow for pre-processing models. They could just strip out the required/default traits if the clientOptional trait is present.